### ISSUE TITLE
Upgade API version to V3

### ISSUE THREAD

**Issue by nguyentoanit** [ScaleLeap/amazon-advertising-api-sdk#399]

**Comment by nguyentoanit** [comment]
@moltar: Could i migrate API to V3 after finish all issue?
At the moment, our lib still requests to API V2. 
If upgrade, we need to update operation and their tests.

**Comment by moltar** [comment]
Yeah, sure!

**Comment by nguyentoanit** [comment]
@moltar : I have tried to upgrade to V3 with url paths (`sb`, `v3/sb`, `v3/hsa`) But always return `Resource Not Found Error`. 
It seems to be not available on Sandbox API.
Do we need create an issue to Amazon?

Or i skip this issue and start working on another project?
And resolve them when they're available.

**Comment by moltar** [comment]
@nguyentoanit can you please share the docs where you've seen the v2 -> v3 upgrade being mentioned? Thanks.

**Comment by nguyentoanit** [comment]
@moltar : I found them on [Release Note](https://advertising.amazon.com/API/docs/en-us/info/release-notes). You can find "Version 3 of the Amazon Advertising API is now available" on Release Note.

I tried `sb` url path on [reference documentation](https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi)
It returned a `Resource Not Found Error`. The docs don't mention about `v3/sb`, `v3/hsa` but i still try them (because i think the docs may be not updated yet).

**Comment by moltar** [comment]
I think the `v3` is limited to the following:

> For Sponsored Brands specific elements, the new version of the API introduces the ability to create a Sponsored Brand campaign via the API, upload and retrieve creative assets, retrieve the brand associated with a specific profile, and access creative and landing page information in the Sponsored Brands campaign object.

**Comment by moltar** [comment]
So maybe try:

 * Sponsored Brand campaign
 * upload and retrieve creative assets
 * retrieve the brand associated with a specific profile

operations

**Comment by nguyentoanit** [comment]
- [create a Sponsored Brand campaign](https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi#/Campaigns/createCampaigns)
- [retrieve the brand associated with a specific profile](https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi#/Brands/getBrands)

I have tried but still get an error.

> upload and retrieve creative assets
I couldn't determine which operation do it.

**Comment by nguyentoanit** [comment]
@moltar : I still have left issue and also reported the result about API V3 on test API above. 👍

**Comment by moltar** [comment]
Hm, which country are you testing this on?

I think some of these new API endpoints only work in US for now.

**Comment by nguyentoanit** [comment]
@moltar : I'm testing on Sandbox API. I don't have a Production token to test.

**Comment by moltar** [comment]
Right, but you can still target different countries based by supplying that to an http constructor.

How about this. Create a failing test PR and I’ll take a look.

**Comment by nguyentoanit** [comment]
@moltar: I created a [failing test](https://github.com/ScaleLeap/amazon-advertising-api-sdk/pull/413). Please help me take a look it. Thanks.

**Comment by moltar** [comment]
From Amazon support:

> Sandbox environment for Sponsored Brands campaigns is not currently available for testing. This is in our backlog but not prioritized at the moment. It is recommended to test the functionality on existing campaigns such as list operations.

So, I think it is not possible to use sandbox for testing.

In which case, we should just ignore tests for now, and just assume that it works 🤷‍♂️

**Comment by MIT120** [comment]
Hi, any Idea if the sdk will migrate to v3 any time soon?
The support for v2 is till march for and June this year.
Many thanks for your time!

**Comment by moltar** [comment]
@MIT120 There are no plans yet. We have to see how the OpenAPI specs shape up over time. We might use the same approach as we use on the SP API SDK to auto-generate the models from the spec. This would be a one-time, breaking change.

**Comment by NiciusB** [comment]
The v2 endpoints for sponsored product reports are being deprecated next week.

If someone needs a solution for now, I've cobbled together a custom operation you can use for V3's Reports:

```
import { Operation } from '@scaleleap/amazon-advertising-api-sdk'
import assert from 'node:assert'

export default class ReportsV3Operation extends Operation {
  async getReport(reportId: string): Promise<AsyncReport> {
    return await this.client.get(`reporting/reports/${reportId}`)
  }

  async deleteReport(reportId: string): Promise<DeleteAsyncReportResponse> {
    return await this.client.delete(`reporting/reports/${reportId}`)
  }

  async createReport(body: CreateAsyncReportRequest): Promise<AsyncReport> {
    return await this.client.post(`reporting/reports`, body)
  }

  async downloadReportDocument(report: AsyncReport) {
    /*
    You'll need to implement this by yourself. I've done something like this:
    
    import { ungzip } from 'node-gzip'
    const fetchRes = await fetch(url)
    const resultText = await ungzip(await fetchRes.arrayBuffer()).then((x) =>
      x.toString('utf-8')
    )
    return JSON.parse(resultText)
   */
  }
}

/**
 * Types created with openapi-to-ts, based on https://dtrnk0o2zy01c.cloudfront.net/openapi/en-us/dest/OfflineReport_prod_3p.json
 */

export type CreateAsyncReportRequest = {
  endDate: string
  configuration: AsyncReportConfiguration
  name?: string
  startDate: string
}

export type AsyncReport = {
  reportId: string
  endDate: string
  configuration: AsyncReportConfiguration
  urlExpiresAt?: string | null
  url?: string | null
  createdAt: string
  fileSize?: number | null
  failureReason?: string | null
  name?: string | null
  generatedAt?: string | null
  startDate: string
  status: 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED'
  updatedAt: string
}

export type AsyncReportFilter = {
  field?: string
  values?: string[]
}

/**
 * The Error Response.
 */
export type AsyncReportingError = {
  code?: string
  details?: string
}

export type DeleteAsyncReportResponse = {
  code?: string
  reportId?: string
  details?: string
}

export type AsyncReportConfiguration = {
  adProduct: AsyncReportAdProduct
  columns: string[]
  reportTypeId: string
  format: 'GZIP_JSON'
  groupBy: string[]
  filters?: AsyncReportFilter[] | null
  timeUnit: 'SUMMARY' | 'DAILY'
}

/**
 * The advertising product such as SPONSORED_PRODUCTS or SPONSORED_BRANDS.
 */
export type AsyncReportAdProduct = 'SPONSORED_PRODUCTS' | 'SPONSORED_BRANDS'
```

**Comment by NiciusB** [comment]
I started adapting it still needs work: https://github.com/NiciusB/amazon-advertising-api-sdk/commit/d6869a596a0ee658148b6874668370652c0a964f

If I end up finishing it I'll open a PR

**Comment by wuqingabc** [comment]
@NiciusB Do you have it done? Need your help on this my friend :-)

**Comment by hung-datadive** [comment]
Do we have any update on V3? Amazon will remove a lot of V2 SP APIs in two weeks.
https://advertising.amazon.com/API/docs/en-us/info/deprecations
<img width="784" alt="image" src="https://github.com/ScaleLeap/amazon-advertising-api-sdk/assets/123790267/87f771cf-27e6-41b8-b37a-2cec44ab0c5b">

**Comment by moltar** [comment]
Currently, we do not have plans to implement additional functionality for v3.
